src/mame/nintendo/playch10.cpp

```diff
@@ -310,38 +629,848 @@ WRITE_LINE_MEMBER(playch10_state::up8w_w)
+{
+	m_cart_sel = data;
+}
+
+
+/*************************************
+ *
+ *  RP5H01 handling
+ *
+ *************************************/
+
+u8 playch10_state::pc10_prot_r()
+{
+	int data = 0xe7;
+
+	// we only support a single cart connected at slot 0
+	if (m_cart_sel == 0)
+	{
+		data |= (~m_rp5h01->counter_r() << 4) & 0x10;    // D4
+		data |= (m_rp5h01->data_r() << 3) & 0x08;        // D3
+	}
+	return data;
+}
+
+void playch10_state::pc10_prot_w(u8 data)
+{
+	// we only support a single cart connected at slot 0
+	if (m_cart_sel == 0)
+	{
+		m_rp5h01->test_w(data & 0x10);       // D4
+		m_rp5h01->clock_w(data & 0x08);      // D3
+		m_rp5h01->reset_w(~data & 0x01);     // D0
+	}
+}
+
+/*************************************
+ *
+ *  Input Ports
+ *
+ *************************************/
+
+void playch10_state::pc10_in0_w(u8 data)
+{
+	// Toggling bit 0 high then low resets both controllers
+	if (data & 1)
+		return;
+
+	// load up the latches
+	for (int i = 0; i < 2; i++)
+		m_input_latch[i] = m_in[i]->read();
+
+	// apply any masking from the BIOS
+	if (m_cntrl_mask)
+	{
+		// mask out select and start
+		m_input_latch[0] &= ~0x0c;
+	}
+}
+
+u8 playch10_state::pc10_in0_r()
+{
+	int ret = m_input_latch[0] & 1;
+
+	// shift
+	m_input_latch[0] >>= 1;
+
+	// some games expect bit 6 to be set because the last entry on the data bus shows up
+	// in the unused upper 3 bits, so typically a read from $4016 leaves 0x40 there.
+	ret |= 0x40;
+
+	return ret;
+}
+
+u8 playch10_state::pc10_in1_r()
+{
+	int ret = m_input_latch[1] & 1;
+
+	// shift
+	m_input_latch[1] >>= 1;
+
+	// do the gun thing
+	if (m_pc10_gun_controller)
+	{
+		int trigger = m_in[0]->read();
+
+		if (!m_sensor->detect_light(m_gunx->read(), m_guny->read()))
+			ret |= 0x08;
+
+		// now, add the trigger if not masked
+		if (!m_cntrl_mask)
+			ret |= (trigger & 2) << 3;
+	}
+
+	// some games expect bit 6 to be set because the last entry on the data bus shows up
+	// in the unused upper 3 bits, so typically a read from $4016 leaves 0x40 there.
+	ret |= 0x40;
+
+	return ret;
+}
+/*************************************
+ *
+ *  PPU External bus handlers
+ *
+ *************************************/
+
+void playch10_state::pc10_chr_w(offs_t offset, u8 data)
+{
+	int bank = BIT(offset, 10, 3);
+	if (m_chr_page[bank].writable)
+		m_chr_page[bank].chr[offset & 0x3ff] = data;
+}
+
+u8 playch10_state::pc10_chr_r(offs_t offset)
+{
+	int bank = BIT(offset, 10, 3);
+	return m_chr_page[bank].chr[offset & 0x3ff];
+}
+
+void playch10_state::pc10_set_mirroring(int mirroring)
+{
+	int bit = mirroring == PPU_MIRROR_HORZ || mirroring == PPU_MIRROR_HIGH;
+
+	switch (mirroring)
+	{
+		case PPU_MIRROR_LOW:
+		case PPU_MIRROR_HIGH:
+			for (int i = 0; i < 4; i++)
+				m_nt_page[i]->set_entry(bit);
+			break;
+		case PPU_MIRROR_HORZ:
+		case PPU_MIRROR_VERT:
+		case PPU_MIRROR_4SCREEN:
+		default:
+			for (int i = 0; i < 4; i++)
+				m_nt_page[i]->set_entry(BIT(i, bit));
+			break;
+	}
+}
+
+/* SIZE MAPPINGS *\
```

Comment:
Please don’t put backslashes at the ends of lines, and remove them if you see them – they have significance to the preprocessor.

Reply:
Comment removed entirely. It seemed useless and now I know why. It was added 13 years ago when CHR banking was moved out of the PPU internals. They switched the units of the size parameter then, so really the comment is only informative of that moment.